File: templates/vite-react/gonvex/messages.go
package backend

import (
	"time"

	"github.com/gonvex/gonvex/pkg/gonvex"
)

type ListMessagesArgs struct{}

type SendMessageArgs struct {
	Body string `json:"body"`
}

type Message struct {
	ID        string    `json:"id"`
	Body      string    `json:"body"`
	Author    string    `json:"author"`
	CreatedAt time.Time `json:"created_at"`
}

func Register(app *gonvex.App) {
	app.Query("messages.list", ListMessages)
	app.Mutation("messages.send", SendMessage)
}

func ListMessages(ctx *gonvex.QueryCtx, args ListMessagesArgs) ([]Message, error) {
	return []Message{}, nil
}

func SendMessage(ctx *gonvex.MutationCtx, args SendMessageArgs) (Message, error) {
	return Message{
		ID:        "pending",
		Body:      args.Body,
		Author:    "demo-user",
		CreatedAt: time.Now().UTC(),
	}, nil
}

File: templates/vite-react/gonvex/messages.ts
import { liveQuery, reducer, schema, visibility } from "@gonvex/module-sdk";

export const messagesVisibility = visibility({
  table: "messages",
  key: "id",
  sets: {},
  where: { operator: "public" },
});

export const list = liveQuery({
  args: schema.object({}),
  result: schema.array(schema.object({
    id: schema.id("messages"),
    body: schema.string(),
    author: schema.string(),
    created_at: schema.datetime(),
  })),
  liveQueryPlan: {
    table: "messages",
    key: "id",
    columns: ["id", "body", "author", "created_at"],
    sort: {
      defaultColumn: "created_at",
      defaultDirection: "asc",
      allowedColumns: ["created_at"],
    },
  },
  run: async () => [],
});

export const send = reducer({
  args: schema.object({ body: schema.string() }),
  result: schema.object({
    id: schema.id("messages"),
    body: schema.string(),
    author: schema.string(),
    created_at: schema.datetime(),
  }),
  offline: { mode: "onlineOnly", reason: "server assigns the message id" },
  nonOptimisticReason: "server assigns the message id",
  run: async ({ now }, args) => ({
    id: "pending",
    body: args.body,
    author: "demo-user",
    created_at: new Date(now).toISOString(),
  }),
});
